### ISSUE TITLE
Multiple where-conditions won't work

### ISSUE THREAD

**Issue by CinKon** [revolver-app/vuex-orm#34]
Happy new year! 🎉 

Unfortunately, when I try to get entities like this:

```javascript
getters['entities/user/query']()
          .where('roleId', 3)
          .where('firstname', (value) => value.includes(searchTerm))
          .orWhere('surname', (value) => value.includes(searchTerm))
          .orWhere('username', (value) => value.includes(searchTerm))
          .orWhere('email', (value) => value.includes(searchTerm))
          .get()
```

`where('roleId', 3)` is being ignored.

**Comment by kiaking** [comment]
Happy new year! 🎉 

And what!? OK I'll look into this 👨‍🎤

**Comment by kiaking** [comment]
@CinKon OK, I'm trying to reproduce this but not sure how... Could you give me the example of the data that `where('roleId', 3)` is being ignored?

I've created test case something like this but it works as expected.

```js
const state = {
  name: 'entities',
  users: { data: {
    '1': { id: 1, name: 'John', role: 'user', age: 20 },
    '2': { id: 2, name: 'Jane', role: 'user', age: 30 },
    '3': { id: 3, name: 'Johnny', role: 'admin', age: 40 }
  }}
}

const expected = [
  { id: 1, name: 'John', role: 'user', age: 20 },
  { id: 2, name: 'Jane', role: 'user', age: 30 },
  { id: 3, name: 'Johnny', role: 'admin', age: 40 }
]

const result = Repo.query(state, 'users', false)
  .where('id', 1)
  .where('name', 'John')
  .orWhere('role', 'admin')
  .orWhere('age', 30)
  .get()

expect(result).toEqual(expected)
```

**Comment by CinKon** [comment]
Ok, looks like we understand this differently.
When I put multiple `where`-Conditions in a row, I'd expect to act them like `WHERE foo = 'bar' AND foo2 = 'bar2'`, but your example makes no difference for me to the `orWhere`-Conditions.

This would be a example I'd expect:

```javascript
const state = {
  name: 'entities',
  users: { data: {
    '1': { id: 1, name: 'John', role: 'user', age: 40 },
    '2': { id: 2, name: 'Jane', role: 'user', age: 30 },
    '3': { id: 3, name: 'Johnny', role: 'admin', age: 20 }
  }}
}

const expected = [
  { id: 1, name: 'John', role: 'user', age: 40 }
]

const result = Repo.query(state, 'users', false)
  .where('role', 'user')
  .where('age', (value) => value > 30)
  .get()

expect(result).toEqual(expected)
```

Since the both conditions are chained with `AND`, both of them have to apply.

Point is, I'm missing an `andWhere`-Function ;)

**Comment by kiaking** [comment]
Thanks for the info! Thats is exactly how things should work I say.

I'm trying to reproduce this issue but somehow I can't. The example code of yours actually works. I've added the test here https://github.com/revolver-app/vuex-orm/pull/36/files#diff-32780d4d8d0fa4cfd793aea28325a424R184 but it's passing 😢 

Maybe it has to do with `orWhere` condition. When `orWhere` is set, the prior `where` will be ignored if the orWhere matches. Because if the `orWhere` condition matches, then that record will be fetched regardless of the where conditions.

I admit that having many `where`s and `orWhere`s can make things complex, especially when they are mixed. In such case maybe using closure would make things clearer.

```js
// Instead of this.
getters['entities/user/query']()
  .where('roleId', 3)
  .where('firstname', (value) => value.includes(searchTerm))
  .orWhere('surname', (value) => value.includes(searchTerm))
  .orWhere('username', (value) => value.includes(searchTerm))
  .orWhere('email', (value) => value.includes(searchTerm))
  .get()

// Maybe this is more understandable.
getters['entities/user/query']().where((user) => {
  return (user.roleId === 3 && user.firstName)
         || user.surname.includes(searchTerm)
         || user.username.includes(searchTerm)
         || user.email.includes(searchTerm)
}).get()
```

At least, the above example works equivalent to each other. Maybe you are expecting something else...?

**Comment by CinKon** [comment]
Yeah, i was afraid that this would be the case...
Is it possible that you could make it so that the `orWhere` only effects the last appearing `where`-Function?
So that 

```javascript
getters['entities/user/query']()
  .where('roleId', 3)
  .where('firstname', (value) => value.includes(searchTerm))
  .orWhere('surname', (value) => value.includes(searchTerm))
  .orWhere('username', (value) => value.includes(searchTerm))
  .orWhere('email', (value) => value.includes(searchTerm))
  .get()
```

results in 

```sql
WHERE roleId = 3 AND (firstname = '%searchtearm%' OR surname = '%searchtearm%' OR username = '%searchtearm%' OR email = '%searchtearm%')
```

Your last example should look like this to me:

```javascript
getters['entities/user/query']().where((user) => {
  return (user.roleId === 3 && (
         user.firstName.includes(searchTerm)
         || user.surname.includes(searchTerm)
         || user.username.includes(searchTerm)
         || user.email.includes(searchTerm)
  ))
}).get()
```

**Comment by kiaking** [comment]
Ah Now I see.

It is possible to make `orWhere` only effects the last appearing `where`, but I don't think that's the way to go about it. Because you could add `where` at any time and the order of `where` might not be always guarantied. In fact, `where` is used internally when resolving relationship, so I think there will be some case you would end up getting unexpected result.

I've checked other ORM framework and found that they solve this issue by nesting the query. For example in Laravel, you can do something like this (doc can be found here  https://laravel.com/docs/5.5/queries#parameter-grouping).

```js
// Translating Laravel (PHP) to JS here.
DB.table('users')
  .where('name', '=', 'John')
  .orWhere((query) => {
    query.where('votes', '>', 100).where('title', '<>', 'Admin')
  })
  .get()
```

Now, following above syntax in Vuex ORM would be little hard since Vuex ORM already have same syntax with passing record object instead of `query` object.

However, passing query object is not necessary for Vuex ORM since Vuex ORM is working against JSON, not SQL based DB. So for your need, I think below example would solve the issue.

```js
getters['entities/user/query']()
  .where('roleId', 3)
  .where((user) => {
    return user.firstName.includes(searchTerm)
           || user.surname.includes(searchTerm)
           || user.username.includes(searchTerm)
           || user.email.includes(searchTerm)
  })
  .get()
```

To "group" conditions together, I think the only way to do that is by using closure. Maybe we can pass `query` object as second argument and do something like this too.

```js
getters['entities/user/query']()
  .where('roleId', 3)
  .where((_user, query) => {
    query.where('firstname', (value) => value.includes(searchTerm))
      .orWhere('surname', (value) => value.includes(searchTerm))
      .orWhere('username', (value) => value.includes(searchTerm))
      .orWhere('email', (value) => value.includes(searchTerm))
  })
  .get()
```

Not sure if this syntax is better though... What do you think?

**Comment by CinKon** [comment]
I think your last example would be more consistent, but I'm fine with both, as long as they work 😄

**Comment by kiaking** [comment]
OK great, I'll implement the nested query clause. Please workaround your issue with former example for now (it works with current version).

**Comment by kiaking** [comment]
I've implemented the feature in Vuex ORM 0.14.1! 🎉 

Now, when passing closure to the first argument, it will also receive query builder as second argument. By using the query builder, you may nest the where clause. This is useful when you want "group" the where clauses.

```js
// Retrieve all users with role of user, and age of 20 or id of 1.
const user = store.getters['entities/users/query']()
  .where('role', 'admin')
  .where((_user, query) => {
    query.where('age', 20).orWhere('id', 1)
  })
  .get()
```

I wasn't thinking much but this resulted in enabling unlimited nested queries! (I wouldn't recommend nesting too deeply though... lol)

Thank you for the great feedback and advise as always!
